ADR/0028-component-dependencies.md

```diff
@@ -0,0 +1,179 @@
+  * In this case, the bundle image must be rebuilt for the operand image update to be tested at all (assuming OLM is the strategy for deploying the operator).
+  * This is a one-to-many dependency. One operator bundle depends on many operand images.
+* A user team just has **two components that depend on each other in a functional way**. When they propose an update to one component, they want to submit a corresponding change to the second component, and have those tested together before merging both.
+  * In this case, on any given day, component B’s PRs may or may not have dependencies on PRs from component A. Not all changes are linked.
+
+In theory **any combination of the cases above** could be present in an app. They could have one common parent image, many operand images that depend on that, and one or more operator bundles which depend on those.
+
+Today, users work around how complicated it is to manage digests themselves by instead using *floating tags*, which have the benefit of being easy to use - no need to update - but have the problem of being unclear. It's not exactly clear what you're building against if you refer to it by tag. Potentially insecure. We want to make it easy for users to do better.
+
+## Decision
+
+### Interface changes
+
+**Declared Component Dependencies**: Introduce a new field on the `Component` resource called `depends on` that lets one Component declare that another Component **depends on it**.
+
+* The `depends on` field is a list.
+* We will call a Component which has a dependency on another Component: a **"dependant"** component.
+* We will call a Component on which other Components have a dependency: a **"dependency"** component.
+* [integration-service] will handle the testing and promotion of all **dependancy** components in a special way.
+* [build-service] will use the `depends on` links to propose updates to users' **dependant** git repositories.
+* The set of Components and their "depends on" links form a directed-acyclic-graph (DAG). Application-service is responsible for making sure that cycles are not introduced and for managing references in the lifecycle of all Components in the Application. Deleting a Component should be well-defined.
+* As with Components in general, the "depends on" references are Application-scoped. Declared relationships between Components in different Applications are not to be considered valid. Declared relationships between Components in different namespaces are not to be considered valid.
+
+**PR Groups**: Introduce a new convention for PRs that lets one PR declare that it is related to another PR.
+
+* Treat the **source branch name** of the PR as a slash-delimited string, and use the first element of that
```

Comment:
Agreed on the first part -  we should revise this to include some scoping by source repo path.

On the second part, I've seen that too, notably in the [gitflow](https://nvie.com/posts/a-successful-git-branching-model/) branching model. To make matters worse, there are two aspects of the design here: Component "depends on" references and PR Groups. The Component depends on references have to be explicitly supplied by the user, but the PR Groups are just assumed to be in use by the system. An unaware developer would have their "feature/foo" and "feature/bar" features tested together in a group.

Reply:
Brainstorming here: what if we introduce a prefix to that branch name that indicates that the originating party intends for it to be tested as part of a group.  Say, `group/...`? Then, if you submit `feature/foo` and `feature/bar` they are tested and handled separately. If you submit `group/foo/part-1` and `group/foo/part-2`, then they are tested as part of the same PR-group "foo".

Reply:
.. no, that won't work. We depend on the ability to submit an automated PR (a digest bump) that can group itself with a previously-submitted manual PR (the originating change). This would only work if the user knew to prefix the branch of their originating change with `group/` - but of course, they don't know that and shouldn't have to know that.

Reply:
It only sidesteps the problem - but we could pick a less commonly used character than `/`. Say, `@`? Where branches in a PR group would look like `feature/foo@part-1` and `feature/foo@part-2` and the PR group name would be taken as `feature/foo`.

Reply:
> - but of course, they don't know that and shouldn't have to know that.

Is this not something that we can overcome with documentation? This will be a feature for AppStudio that users would be able to _discover_ and then utilize. If users want to take advantage of this after the fact then they can change the source branch for the PR.

Reply:
> I think it is concerning that this is not scoped to developers in any way, so if I have stuartwdouglas/minor and someone else has a similar otherdev/minor branch this will consider them to be part of the same PR group.

On second thought, I think we actually want this, so that a user can submit a pr from their fork with a certain branch name, and our system can respond with a followup PR from _its_ fork using the same branch name in order to induce combined testing for a PR group.

Reply:
Updated in cf1c549.